spec/compiler/crystal/tools/init_spec.cr

```diff
@@ -1,149 +1,175 @@
-require "spec"
-require "yaml"
-require "ini"
 require "compiler/crystal/config"
 require "compiler/crystal/tools/init"
+require "file_utils"
+require "ini"
+require "spec"
+require "yaml"
+
+BIN_CRYSTAL = File.expand_path("bin/crystal")
+
+private def exec_init(project_name, project_dir = nil, type = "lib")
+  args = ["init", type, project_name]
+  args << project_dir if project_dir
+
+  process = Process.new(BIN_CRYSTAL, args, shell: true, error: Process::Redirect::Pipe)
+  stderr = process.error.gets_to_end
+  status = process.wait
+  $? = status
+  stderr
+end
+
+# Creates a temporary directory, cd to it and run the block inside it.
+# The directory and its content is deleted when the block return.
+private def within_temporary_directory
+  tmp_path = "#{__DIR__}/crystal_init_spec_tmp_dir-#{rand(10_000)}"
```

Comment:
Now temporary directories will be created relative to the location of each spec file (`__DIR__`).

In that case, will be great to ignore `crystal_init_spec_tmp_dir-*` in `.gitignore` to avoid a broken spec allow possible left over files be available to commit into the codebase.

Reply:
It's a helper for this spec file only, and unless this method `within_temporary_directory` crash, the ensure will automatically clean the tmp dir, so I'm not sure it's necessary

Reply:
The tmp dir should be inside de project root. I think @bcardiff's [comment](https://github.com/crystal-lang/crystal/pull/4915#discussion_r151838447) asking for this was a bit imprecise.
The root folder's `.gitignore` contains `/tmp`, so you should probably use `#{DIR}/../../../../tmp/crystal_init_spec-*`  as temporary directories.

Reply:
@straight-shoota I don't think that's what he meant

Reply:
Currently all these specs output to `/tmp`, which is in `.gitignore`. The change here using `__DIR__` place these temporary files relative to the spec file, inside the `spec` folder, which are not ignored.

> It's a helper for this spec file only, and unless this method within_temporary_directory crash, the ensure will automatically clean the tmp dir, so I'm not sure it's necessary

That is something we don't know. Perhaps `Process.new` might segfault, or other part of the spec could, leaving leftovers that were never cleaned up.

I'm not suggesting radically change and refactor what you have done, but instead consider where those temporary files are stored and either 1: add them to the gitignore file to avoid accidental commit, or 2: adjust the location to be the one relative to the already ignored one.

Thank you.